From foo@baz Wed Oct 26 06:44:13 PM CEST 2022
From: Jiri Olsa <jolsa@kernel.org>
Date: Wed, 19 Oct 2022 10:56:02 +0200
Subject: kbuild: skip per-CPU BTF generation for pahole v1.18-v1.21
To: stable@vger.kernel.org
Cc: "Mel Gorman" <mgorman@techsingularity.net>, "Arnaldo Carvalho de Melo" <acme@redhat.com>, "Hao Luo" <haoluo@google.com>, "Michal Suchanek" <msuchanek@suse.de>, bpf@vger.kernel.org, "Arnaldo Carvalho de Melo" <acme@kernel.org>, "Daniel Borkmann" <daniel@iogearbox.net>, "Martin Rodriguez Reboredo" <yakoyoku@gmail.com>, "Andrii Nakryiko" <andrii@kernel.org>, "Javier Martinez Canillas" <javierm@redhat.com>, "Ilya Leoshkevich" <iii@linux.ibm.com>, "Nathan Chancellor" <nathan@kernel.org>, "Philip Müller" <philm@manjaro.org>
Message-ID: <20221019085604.1017583-4-jolsa@kernel.org>

From: Andrii Nakryiko <andrii@kernel.org>

commit a0b8200d06ad6450c179407baa5f0f52f8cfcc97 upstream.

[small context changes due to missing floats support in 5.10]

Commit "mm/page_alloc: convert per-cpu list protection to local_lock" will
introduce a zero-sized per-CPU variable, which causes pahole to generate
invalid BTF.  Only pahole versions 1.18 through 1.21 are impacted, as
before 1.18 pahole doesn't know anything about per-CPU variables, and 1.22
contains the proper fix for the issue.

Luckily, pahole 1.18 got --skip_encoding_btf_vars option disabling BTF
generation for per-CPU variables in anticipation of some unanticipated
problems.  So use this escape hatch to disable per-CPU var BTF info on
those problematic pahole versions.  Users relying on availability of
per-CPU var BTFs would need to upgrade to pahole 1.22+, but everyone won't
notice any regressions.

Link: https://lkml.kernel.org/r/20210530002536.3193829-1-andrii@kernel.org
Signed-off-by: Andrii Nakryiko <andrii@kernel.org>
Acked-by: Mel Gorman <mgorman@techsingularity.net>
Cc: Arnaldo Carvalho de Melo <acme@redhat.com>
Cc: Hao Luo <haoluo@google.com>
Cc: Michal Suchanek <msuchanek@suse.de>
Cc: Jiri Olsa <jolsa@kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Jiri Olsa <jolsa@kernel.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 scripts/link-vmlinux.sh |    5 +++++
 1 file changed, 5 insertions(+)

--- a/scripts/link-vmlinux.sh
+++ b/scripts/link-vmlinux.sh
@@ -161,6 +161,11 @@ gen_btf()
 
 	vmlinux_link ${1}
 
+	if [ "${pahole_ver}" -ge "118" ] && [ "${pahole_ver}" -le "121" ]; then
+		# pahole 1.18 through 1.21 can't handle zero-sized per-CPU vars
+		extra_paholeopt="${extra_paholeopt} --skip_encoding_btf_vars"
+	fi
+
 	info "BTF" ${2}
 	LLVM_OBJCOPY="${OBJCOPY}" ${PAHOLE} -J ${extra_paholeopt} ${1}
 
